xcresult/src/xcresult.rs

```diff
@@ -0,0 +1,348 @@
+            action
+                .get("startedTime")
+                .and_then(|r| r.get("_value"))
+                .and_then(|r| r.as_str()),
+        ) {
+            let ended_time_parsed = chrono::DateTime::parse_from_str(ended_time, DATE_FORMAT)?;
+            let started_time_parsed = chrono::DateTime::parse_from_str(started_time, DATE_FORMAT)?;
+            let duration = (ended_time_parsed.timestamp_millis()
+                - started_time_parsed.timestamp_millis()) as u64;
+            testsuites_junit.set_time(std::time::Duration::from_millis(duration));
+        }
+        let found_tests = self.find_tests(id)?;
+        let test_summaries = match found_tests.get("summaries").and_then(|r| r.get("_values")) {
+            Some(val) => val.as_array(),
+            None => return Ok(testsuites_junit),
+        };
+        if let Some(test_summaries) = test_summaries {
+            for test_summary in test_summaries {
+                let testable_summaries = match test_summary
+                    .get("testableSummaries")
+                    .and_then(|t| t.get("_values"))
+                    .and_then(|r| r.as_array())
+                {
+                    Some(val) => val,
+                    None => {
+                        return Ok(testsuites_junit);
+                    }
+                };
+                for testable_summary in testable_summaries {
+                    let top_level_tests = match testable_summary
+                        .get("tests")
+                        .and_then(|t| t.get("_values"))
+                        .and_then(|r| r.as_array())
+                    {
+                        Some(val) => val,
+                        None => {
+                            return Ok(testsuites_junit);
+                        }
+                    };
+                    for top_level_test in top_level_tests {
+                        let testsuites = match top_level_test
+                            .get("subtests")
+                            .and_then(|t| t.get("_values"))
+                            .and_then(|r| r.as_array())
+                        {
+                            Some(val) => val,
+                            None => {
+                                return Ok(testsuites_junit);
+                            }
+                        };
+                        for testsuite in testsuites {
+                            let testsuite_junit = self.junit_testsuite(action, testsuite)?;
+                            testsuites_junit.add_test_suite(testsuite_junit);
+                        }
+                    }
+                }
+            }
+        }
+        Ok(testsuites_junit)
+    }
+
+    pub fn generate_junits(&self) -> Result<Vec<Report>, anyhow::Error> {
+        let mut report_junits: Vec<Report> = Vec::new();
+        if let Some(actions) = self
+            .results_obj
+            .get("actions")
+            .and_then(|a| a.get("_values"))
```

Comment:
very optional, might be a future improvement rather than being included in this PR: there seems to be a lot of `&'static str`—like `"_values"`—which we repeat throughout the parsing code. We can make it a bit easier to understand that these are all the same values by making these values constants which we reuse.

Reply:
Maybe I'm misunderstanding, but I don't believe we are grabbing the same values. The structure of the json is such that it's "key" -> "_values"/"_value", which correspond to the underlying value for the particular field. The key is different each time in the above call.

Reply:
`"_value"` might be a better example since it's repeated more often. What I'm proposing is something like

```rs
const RESULTS_FIELD_VALUE: &str: "_value";

// ...
x.get(RESULTS_FIELD_VALUE)
// ...
y.get(RESULTS_FIELD_VALUE)
// ...
z.get(RESULTS_FIELD_VALUE)
```